Fails to start on 5.4

This one fails to start on Mattermost 5.4:

`{"level":"error","ts":1542655298.7106464,"caller":"app/plugin.go:70","msg":"Unable to activate plugin","plugin_id":"com.github.moussetc.mattermost.plugin.emoticon2emoji","error":"unable to start plugin: com.github.moussetc.mattermost.plugin.emoticon2emoji: plugin exited before we could connect","errorVerbose":"plugin exited before we could connect\nunable to start plugin: com.github.moussetc.mattermost.plugin.emoticon2emoji\ngithub.com/mattermost/mattermost-server/plugin.(*Environment).Activate\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/plugin/environment.go:211\ngithub.com/mattermost/mattermost-server/app.(*App).SyncPluginsActiveState\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/app/plugin.go:68\ngithub.com/mattermost/mattermost-server/app.(*App).InitPlugins.func2\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/app/plugin.go:141\ngithub.com/mattermost/mattermost-server/app.(*App).InvokeConfigListeners\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/app/config.go:138\ngithub.com/mattermost/mattermost-server/app.(*App).UpdateConfig\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/app/config.go:50\ngithub.com/mattermost/mattermost-server/app.(*App).EnablePlugin\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/app/plugin.go:206\ngithub.com/mattermost/mattermost-server/api4.enablePlugin\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/api4/plugin.go:183\ngithub.com/mattermost/mattermost-server/web.Handler.ServeHTTP\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/web/handlers.go:131\ngithub.com/mattermost/mattermost-server/web.(*Handler).ServeHTTP\n\t<autogenerated>:1\ngithub.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux.(*Router).ServeHTTP\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/mux/mux.go:162\ngithub.com/mattermost/mattermost-server/vendor/github.com/gorilla/handlers.recoveryHandler.ServeHTTP\n\t/home/ubuntu/jenkins/workspace/msr/mattermost-platform-release/src/github.com/mattermost/mattermost-server/vendor/github.com/gorilla/handlers/recovery.go:78\ngithub.com/mattermost/mattermost-server/vendor/github.com/gorilla/handlers.(*recoveryHandler).ServeHTTP\n\t<autogenerated>:1\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2741\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1847\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"}`

Doesn't compile either:

```
$ make
go test -race -coverprofile=coverage.txt -covermode=atomic
plugin.go:9:2: cannot find package "github.com/mattermost/mattermost-server/plugin/rpcplugin" in any of:
        /usr/lib/go-1.10/src/github.com/mattermost/mattermost-server/plugin/rpcplugin (from $GOROOT)
        /home/chrish/go/src/github.com/mattermost/mattermost-server/plugin/rpcplugin (from $GOPATH)
Makefile:39: recipe for target 'test-coverage' failed
make: *** [test-coverage] Error 1
```

Looks like the code needs to be upgraded for APIv4?

--- a/Gopkg.toml
+++ b/Gopkg.toml
@@ -27,7 +27,7 @@
 
 [[constraint]]
   name = "github.com/mattermost/mattermost-server"
-  version = "5.0.0"
+  version = "~5.5.0"
 
 [prune]
   go-tests = true

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 #
 # Makefile for Mattermost emoticon2emoji
 #
-SRC=plugin.go emoticon.go matches.go
+SRC=plugin.go configuration.go emoticon.go matches.go
 EXEC=plugin
 CONF=plugin.yaml
 PACKAGE_BASENAME=mattermost-plugin-emoticon2emoji

--- a/README.md
+++ b/README.md
@@ -1,4 +1,4 @@
-# mattermost-plugin-emoticon2emoji
+# mattermost-plugin-emoticon2emoji [![Build Status](https://api.travis-ci.com/moussetc/mattermost-plugin-emoticon2emoji.svg?branch=master)](https://travis-ci.com/moussetc/mattermost-plugin-emoticon2emoji)
 A plugin that completes the automatic conversion from emoticon to emoji (:) to :smile:) in Mattermost messages, by adding [Slack mappings](https://get.slack.help/hc/en-us/articles/202931348-Use-emoji-and-emoticons#use-emoticons) and some other (see `matches.go` for the default list) which can be configured.
 
 ## Compatibility

--- a/README.md
+++ b/README.md
@@ -1,8 +1,11 @@
 # mattermost-plugin-emoticon2emoji
 A plugin that completes the automatic conversion from emoticon to emoji (:) to :smile:) in Mattermost messages, by adding [Slack mappings](https://get.slack.help/hc/en-us/articles/202931348-Use-emoji-and-emoticons#use-emoticons) and some other (see `matches.go` for the default list) which can be configured.
 
-## Requirements
-- Mattermost 5.0 (to allow plugins to intercept posts).
+## Compatibility
+- for Mattermost 5.2 or higher: use v2.x.x release
+- for Mattermost 5.0: use v1.0.0 release
+- for Mattermost below: unsupported versions (plugins can't intercept posts)
+
 
 ## Installation and configuration
 1. Go to the [Releases page](https://github.com/moussetc/mattermost-plugin-emoticon2emoji/releases) and download the package for your OS and architecture.

--- a/configuration.go
+++ b/configuration.go
@@ -1,0 +1,88 @@
+package main
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// configuration captures the plugin's external configuration as exposed in the Mattermost server
+// configuration, as well as values computed from the configuration. Any public fields will be
+// deserialized from the Mattermost server configuration in OnConfigurationChange.
+//
+// As plugins are inherently concurrent (hooks being called asynchronously), and the plugin
+// configuration can change at any time, access to the configuration must be synchronized. The
+// strategy used in this plugin is to guard a pointer to the configuration, and clone the entire
+// struct whenever it changes. You may replace this with whatever strategy you choose.
+//
+// If you add non-reference types to your configuration struct, be sure to rewrite Clone as a deep
+// copy appropriate for your types.
+type Emoticon2EmojiPluginConfiguration struct {
+	CustomMatches string
+	matches       map[string]match `yaml:"-",json:"-"`
+}
+
+// Clone shallow copies the configuration. Your implementation may require a deep copy if
+// your configuration has reference types.
+func (c *Emoticon2EmojiPluginConfiguration) Clone() *Emoticon2EmojiPluginConfiguration {
+	var clone = *c
+	return &clone
+}
+
+// getConfiguration retrieves the active configuration under lock, making it safe to use
+// concurrently. The active configuration may change underneath the client of this method, but
+// the struct returned by this API call is considered immutable.
+func (p *Emoticon2EmojiPlugin) getConfiguration() *Emoticon2EmojiPluginConfiguration {
+	p.configurationLock.RLock()
+	defer p.configurationLock.RUnlock()
+
+	if p.configuration == nil {
+		return &Emoticon2EmojiPluginConfiguration{}
+	}
+
+	return p.configuration
+}
+
+// setConfiguration replaces the active configuration under lock.
+//
+// Do not call setConfiguration while holding the configurationLock, as sync.Mutex is not
+// reentrant. In particular, avoid using the plugin API entirely, as this may in turn trigger a
+// hook back into the plugin. If that hook attempts to acquire this lock, a deadlock may occur.
+//
+// This method panics if setConfiguration is called with the existing configuration. This almost
+// certainly means that the configuration was modified without being cloned and may result in
+// an unsafe access.
+func (p *Emoticon2EmojiPlugin) setConfiguration(configuration *Emoticon2EmojiPluginConfiguration) {
+	p.configurationLock.Lock()
+	defer p.configurationLock.Unlock()
+
+	if configuration != nil && p.configuration == configuration {
+		// Ignore assignment if the configuration struct is empty. Go will optimize the
+		// allocation for same to point at the same memory address, breaking the check
+		// above.
+		if reflect.ValueOf(*configuration).NumField() == 0 {
+			return
+		}
+
+		panic("setConfiguration called with the existing configuration")
+	}
+
+	p.configuration = configuration
+}
+
+// OnConfigurationChange is invoked when configuration changes may have been made.
+func (p *Emoticon2EmojiPlugin) OnConfigurationChange() error {
+	var configuration = new(Emoticon2EmojiPluginConfiguration)
+
+	// Load the public configuration fields from the Mattermost server configuration.
+	if err := p.API.LoadPluginConfiguration(configuration); err != nil {
+		return errors.Wrap(err, "failed to load plugin configuration")
+	}
+
+	p.setConfiguration(configuration)
+	if err := p.applyNewConfig(configuration); err != nil {
+		return appError("Unable to apply new plugin configuration", err)
+	}
+	return nil
+}
+

--- a/emoticon.go
+++ b/emoticon.go
@@ -13,11 +13,6 @@ type match struct {
 	regexp      *regexp.Regexp
 }
 
-// Emoticon2EmojiPluginConfiguration is the serialisable plugin configuration
-type Emoticon2EmojiPluginConfiguration struct {
-	CustomMatches string
-}
-
 func (p *Emoticon2EmojiPlugin) applyNewConfig(configuration *Emoticon2EmojiPluginConfiguration) error {
 	// read custom map from config
 	CustomMatches, err := unserializeConfigMatches(configuration.CustomMatches)
@@ -30,9 +25,9 @@ func (p *Emoticon2EmojiPlugin) applyNewConfig(configuration *Emoticon2EmojiPlugi
 	for k, v := range CustomMatches {
 		effectiveMap[k] = v
 	}
-	p.matches = map[string]match{}
+	configuration.matches = map[string]match{}
 	for emoticon, emoji := range effectiveMap {
-		p.matches[emoticon] = match{
+		configuration.matches[emoticon] = match{
 			replacement: emoji,
 			regexp:      getEmoticonRegexp(emoticon),
 		}
@@ -63,8 +58,9 @@ func unserializeConfigMatches(matches string) (map[string]string, error) {
 
 // Translate replace all the configured emoticons in a string by their equivalent as Mattermost emojis
 func (p *Emoticon2EmojiPlugin) translate(input string) (result string) {
-	if p.matches != nil {
-		return translate(input, p.matches)
+	config := p.getConfiguration()
+	if config != nil && config.matches != nil {
+		return translate(input, config.matches)
 	}
 	return input
 }

--- a/plugin.go
+++ b/plugin.go
@@ -2,49 +2,35 @@ package main
 
 import (
 	"net/http"
-	"sync/atomic"
+	"sync"
 
 	"github.com/mattermost/mattermost-server/model"
 	"github.com/mattermost/mattermost-server/plugin"
-	"github.com/mattermost/mattermost-server/plugin/rpcplugin"
 )
 
 // Emoticon2EmojiPlugin is a Mattermost plugin that replace text emoticons in messages by an emoji approximation
 type Emoticon2EmojiPlugin struct {
-	api           plugin.API
-	configuration atomic.Value
-	matches       map[string]match
+	plugin.MattermostPlugin
+	// configurationLock synchronizes access to the configuration.
+	configurationLock sync.RWMutex
+	// configuration is the active plugin configuration. Consult getConfiguration and
+	// setConfiguration for usage.
+	configuration *Emoticon2EmojiPluginConfiguration
 }
 
 // OnActivate register the plugin command
-func (p *Emoticon2EmojiPlugin) OnActivate(api plugin.API) error {
-	p.api = api
-
+func (p *Emoticon2EmojiPlugin) OnActivate() error {
 	return p.OnConfigurationChange()
 }
 
-// OnConfigurationChange apply a new plugin configuration
-func (p *Emoticon2EmojiPlugin) OnConfigurationChange() error {
-	var configuration Emoticon2EmojiPluginConfiguration
-	if err := p.api.LoadPluginConfiguration(&configuration); err != nil {
-		return appError("Unable to load new plugin configuration", err)
-	}
-
-	p.configuration.Store(&configuration)
-	if err := p.applyNewConfig(&configuration); err != nil {
-		return appError("Unable to apply new plugin configuration", err)
-	}
-	return nil
-}
-
 // MessageWillBePosted converts emoticons in new posts
-func (p *Emoticon2EmojiPlugin) MessageWillBePosted(post *model.Post) (*model.Post, string) {
+func (p *Emoticon2EmojiPlugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*model.Post, string) {
 	post.Message = p.translate(post.Message)
 	return post, ""
 }
 
 // MessageWillBeUpdated converts emoticons in edited posts
-func (p *Emoticon2EmojiPlugin) MessageWillBeUpdated(newPost, oldPost *model.Post) (*model.Post, string) {
+func (p *Emoticon2EmojiPlugin) MessageWillBeUpdated(c *plugin.Context, newPost, oldPost *model.Post) (*model.Post, string) {
 	newPost.Message = p.translate(newPost.Message)
 	return newPost, ""
 }
@@ -59,5 +45,5 @@ func appError(message string, err error) *model.AppError {
 
 // Install the RCP plugin
 func main() {
-	rpcplugin.Main(&Emoticon2EmojiPlugin{})
+	plugin.ClientMain(&Emoticon2EmojiPlugin{})
 }

--- a/plugin.yaml
+++ b/plugin.yaml
@@ -1,7 +1,7 @@
 id: com.github.moussetc.mattermost.plugin.emoticon2emoji
 name: Emoticon to Emoji
 description: "Add Slack emoticon to emoji conversion and some others (like XD to :laughing:) and allow customized mappings"
-version: '1.0.0'
+version: '2.0.0'
 backend:
     executable: plugin
 settings_schema:

--- a/plugin_test.go
+++ b/plugin_test.go
@@ -13,10 +13,13 @@ import (
 
 func initTestPlugin(t *testing.T, config *Emoticon2EmojiPluginConfiguration) *plugintest.API {
 	api := &plugintest.API{}
-	api.On("LoadPluginConfiguration", mock.MatchedBy(func(x interface{}) bool { return true })).Run(func(args mock.Arguments) {
+	api.On("LoadPluginConfiguration", mock.AnythingOfType("*main.Emoticon2EmojiPluginConfiguration")).Run(func(args mock.Arguments) {
 		dest := args.Get(0).(*Emoticon2EmojiPluginConfiguration)
 		dest.CustomMatches = config.CustomMatches
-	}).Return(nil)
+	}).Return(func(dest interface{}) error {
+		*dest.(*Emoticon2EmojiPluginConfiguration) = *config
+		return nil
+	})
 	return api
 }
 
@@ -25,12 +28,13 @@ func TestNewPostWithDefaultEmoticon(t *testing.T) {
 	config := Emoticon2EmojiPluginConfiguration{
 		CustomMatches: "{\"XD\":\"laughing\"}",
 	}
-	p.OnActivate(initTestPlugin(t, &config))
+	p.API = initTestPlugin(t, &config)
+	p.OnConfigurationChange()
 
 	post := &model.Post{
 		Message: "Hello XD !!",
 	}
-	post, err := p.MessageWillBePosted(post)
+	post, err := p.MessageWillBePosted(nil, post)
 	assert.NotNil(t, post)
 	assert.Equal(t, "Hello :laughing: !!", post.Message)
 	assert.Equal(t, "", err)
@@ -39,12 +43,13 @@ func TestNewPostWithDefaultEmoticon(t *testing.T) {
 func TestUpdatedPostWithSlackEmoticon(t *testing.T) {
 	p := Emoticon2EmojiPlugin{}
 	config := Emoticon2EmojiPluginConfiguration{}
-	p.OnActivate(initTestPlugin(t, &config))
+	p.API = initTestPlugin(t, &config)
+	p.OnConfigurationChange()
 
 	post := &model.Post{
 		Message: "Hello </3 !!",
 	}
-	post, err := p.MessageWillBeUpdated(post, nil)
+	post, err := p.MessageWillBeUpdated(nil, post, nil)
 	assert.NotNil(t, post)
 	assert.Equal(t, "Hello :broken_heart: !!", post.Message)
 	assert.Equal(t, "", err)
@@ -96,12 +101,13 @@ func TestMappingsPrecedence(t *testing.T) {
 	config := Emoticon2EmojiPluginConfiguration{
 		CustomMatches: "{\"8)\":\"cry\"}",
 	}
-	p.OnActivate(initTestPlugin(t, &config))
+	p.API = initTestPlugin(t, &config)
+	p.OnConfigurationChange()
 
 	post := &model.Post{
 		Message: "8)",
 	}
-	post, err := p.MessageWillBeUpdated(post, nil)
+	post, err := p.MessageWillBeUpdated(nil, post, nil)
 	assert.NotNil(t, post)
 	assert.Equal(t, ":cry:", post.Message)
 	assert.Equal(t, "", err)

